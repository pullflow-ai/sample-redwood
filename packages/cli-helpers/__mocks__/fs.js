import * as memfs from 'memfs'

export * from 'memfs'
export default memfs.fs
