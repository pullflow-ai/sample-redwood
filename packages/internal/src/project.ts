import fs from 'fs'
import path from 'path'

import { parseConfigFileTextToJson } from 'typescript'

import { getPaths } from '@redwoodjs/project-config'

export const getTsConfigs = () => {
  const rwPaths = getPaths()
  const apiTsConfigPath = path.join(rwPaths.api.base, 'tsconfig.json')
  const webTsConfigPath = path.join(rwPaths.web.base, 'tsconfig.json')

  const apiTsConfig = fs.existsSync(apiTsConfigPath)
    ? parseConfigFileTextToJson(
        apiTsConfigPath,
        fs.readFileSync(apiTsConfigPath, 'utf-8'),
      )
    : null

  const webTsConfig = fs.existsSync(webTsConfigPath)
    ? parseConfigFileTextToJson(
        webTsConfigPath,
        fs.readFileSync(webTsConfigPath, 'utf-8'),
      )
    : null

  return {
    api: apiTsConfig?.config ?? null,
    web: webTsConfig?.config ?? null,
  }
}

export const isTypeScriptProject = () => {
  const paths = getPaths()
  return (
    fs.existsSync(path.join(paths.web.base, 'tsconfig.json')) ||
    fs.existsSync(path.join(paths.api.base, 'tsconfig.json'))
  )
}

export const isServerFileSetup = () => {
  const serverFilePath = path.join(
    getPaths().api.src,
    `server.${isTypeScriptProject() ? 'ts' : 'js'}`,
  )

  return fs.existsSync(serverFilePath)
}

export const isRealtimeSetup = () => {
  const realtimePath = path.join(
    getPaths().api.lib,
    `realtime.${isTypeScriptProject() ? 'ts' : 'js'}`,
  )

  return fs.existsSync(realtimePath)
}
