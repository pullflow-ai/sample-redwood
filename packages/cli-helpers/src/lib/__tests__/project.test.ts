vi.mock('fs')
vi.mock('node:fs', async () => {
  const memfs = await import('memfs')
  return {
    ...memfs.fs,
    default: memfs.fs,
  }
})

import * as fs from 'node:fs'

import * as toml from 'smol-toml'
import { vi, describe, beforeEach, afterEach, it, expect } from 'vitest'

import { updateTomlConfig, addEnvVar } from '../project.js'

const defaultRedwoodToml: Record<string, any> = {
  web: {
    title: 'Redwood App',
    port: 8910,
    apiUrl: '/.redwood/functions',
    includeEnvironmentVariables: [],
  },
  api: {
    port: 8911,
  },
}

const getRedwoodToml = () => {
  return defaultRedwoodToml
}

vi.mock('@redwoodjs/project-config', () => {
  return {
    getPaths: () => {
      return {
        generated: {
          base: '.redwood',
        },
        base: '',
      }
    },
    getConfigPath: () => {
      return '.redwood.toml'
    },
    getConfig: () => {
      return getRedwoodToml()
    },
  }
})

describe('addEnvVar', () => {
  let envFileContent = ''

  describe('addEnvVar adds environment variables as part of a setup task', () => {
    beforeEach(() => {
      vi.spyOn(fs, 'existsSync').mockReturnValue(true)

      vi.spyOn(fs, 'readFileSync').mockImplementation(() => envFileContent)

      vi.spyOn(fs, 'writeFileSync').mockImplementation((envPath, envFile) => {
        expect(envPath).toContain('.env')
        return envFile
      })
    })

    afterEach(() => {
      vi.restoreAllMocks()
      envFileContent = ''
    })

    it('should add a new environment variable when it does not exist', () => {
      envFileContent = 'EXISTING_VAR = value\n# CommentedVar = 123\n'
      const file = addEnvVar('NEW_VAR', 'new_value', 'New Variable Comment')

      expect(file).toMatchSnapshot()
    })

    it('should add a new environment variable when it does not exist when existing envars have no spacing', () => {
      envFileContent = 'EXISTING_VAR=value\n# CommentedVar = 123\n'
      const file = addEnvVar('NEW_VAR', 'new_value', 'New Variable Comment')

      expect(file).toMatchSnapshot()
    })

    it('should add a comment that the existing environment variable value was not changed, but include its new value as a comment', () => {
      envFileContent = 'EXISTING_VAR=value\n# CommentedVar=123\n'
      const file = addEnvVar(
        'EXISTING_VAR',
        'new_value',
        'Updated existing variable Comment',
      )

      expect(file).toMatchSnapshot()
    })

    it('should handle existing environment variables with quoted values', () => {
      envFileContent = `EXISTING_VAR = "value"\n# CommentedVar = 123\n`
      const file = addEnvVar('EXISTING_VAR', 'value', 'New Variable Comment')

      expect(file).toMatchSnapshot()
    })

    it('should handle existing environment variables with quoted values and no spacing', () => {
      envFileContent = `EXISTING_VAR="value"\n# CommentedVar=123\n`
      const file = addEnvVar('EXISTING_VAR', 'value', 'New Variable Comment')

      expect(file).toMatchSnapshot()
    })

    it('should handle existing environment variables and new value with quoted values by not updating the original value', () => {
      envFileContent = `EXISTING_VAR = "value"\n# CommentedVar = 123\n`
      const file = addEnvVar(
        'EXISTING_VAR',
        'new_value',
        'New Variable Comment',
      )

      expect(file).toMatchSnapshot()
    })
  })
})

describe('updateTomlConfig', () => {
  describe('updateTomlConfig configures a new CLI plugin', () => {
    beforeEach(() => {
      vi.spyOn(fs, 'existsSync').mockReturnValue(true)

      vi.spyOn(fs, 'readFileSync').mockImplementation(() => {
        return toml.stringify(defaultRedwoodToml)
      })

      vi.spyOn(fs, 'writeFileSync').mockImplementation((tomlPath, tomlFile) => {
        expect(tomlPath).toContain('redwood.toml')
        return tomlFile
      })
    })

    afterEach(() => {
      vi.restoreAllMocks()
    })

    it('adds when experimental cli is not configured', () => {
      const file = updateTomlConfig(
        '@example/test-package-when-cli-not-configured',
      )
      expect(file).toMatchSnapshot()
    })

    it('adds when experimental cli has some plugins configured', () => {
      defaultRedwoodToml.experimental = {
        cli: {
          autoInstall: true,
          plugins: [
            {
              package:
                '@existing-example/some-package-when-cli-has-some-packages-configured',
            },
          ],
        },
      }

      const file = updateTomlConfig('@example/test-package-name')
      expect(file).toMatchSnapshot()
    })

    it('adds when experimental cli is setup but has no plugins configured', () => {
      defaultRedwoodToml.experimental = {
        cli: {
          autoInstall: true,
        },
      }

      const file = updateTomlConfig(
        '@example/test-package-when-no-plugins-configured',
      )

      expect(file).toMatchSnapshot()
    })

    it('adds package but keeps autoInstall false', () => {
      defaultRedwoodToml.experimental = {
        cli: {
          autoInstall: false,
        },
      }

      const file = updateTomlConfig(
        '@example/test-package-when-autoInstall-false',
      )

      expect(file).toMatchSnapshot()
    })

    it('does not add duplicate place when experimental cli has that plugin configured', () => {
      defaultRedwoodToml.experimental = {
        cli: {
          autoInstall: true,
          plugins: [
            {
              package: '@existing-example/some-package-name-already-exists',
            },
          ],
        },
      }

      const file = updateTomlConfig(
        '@existing-example/some-package-name-already-exists',
      )

      expect(file).toMatchSnapshot()
    })
  })
})
