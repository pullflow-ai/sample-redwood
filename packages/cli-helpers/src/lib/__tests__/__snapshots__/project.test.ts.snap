// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`addEnvVar > addEnvVar adds environment variables as part of a setup task > should add a comment that the existing environment variable value was not changed, but include its new value as a comment 1`] = `
"EXISTING_VAR=value
# CommentedVar=123

# Note: The existing environment variable EXISTING_VAR was not overwritten. Uncomment to use its new value.
# Updated existing variable Comment
# EXISTING_VAR=new_value
"
`;

exports[`addEnvVar > addEnvVar adds environment variables as part of a setup task > should add a new environment variable when it does not exist 1`] = `
"EXISTING_VAR = value
# CommentedVar = 123

# New Variable Comment
NEW_VAR=new_value
"
`;

exports[`addEnvVar > addEnvVar adds environment variables as part of a setup task > should add a new environment variable when it does not exist when existing envars have no spacing 1`] = `
"EXISTING_VAR=value
# CommentedVar = 123

# New Variable Comment
NEW_VAR=new_value
"
`;

exports[`addEnvVar > addEnvVar adds environment variables as part of a setup task > should handle existing environment variables and new value with quoted values by not updating the original value 1`] = `
"EXISTING_VAR = "value"
# CommentedVar = 123

# Note: The existing environment variable EXISTING_VAR was not overwritten. Uncomment to use its new value.
# New Variable Comment
# EXISTING_VAR=new_value
"
`;

exports[`addEnvVar > addEnvVar adds environment variables as part of a setup task > should handle existing environment variables with quoted values 1`] = `
"EXISTING_VAR = "value"
# CommentedVar = 123
"
`;

exports[`addEnvVar > addEnvVar adds environment variables as part of a setup task > should handle existing environment variables with quoted values and no spacing 1`] = `
"EXISTING_VAR="value"
# CommentedVar=123
"
`;

exports[`updateTomlConfig > updateTomlConfig configures a new CLI plugin > adds package but keeps autoInstall false 1`] = `
"[web]
title = "Redwood App"
port = 8910
apiUrl = "/.redwood/functions"
includeEnvironmentVariables = []

[api]
port = 8911

[experimental]
[experimental.cli]
autoInstall = false
[experimental]
[experimental.cli]
[[experimental.cli.plugins]]
package = "@example/test-package-when-autoInstall-false"
enabled = true
"
`;

exports[`updateTomlConfig > updateTomlConfig configures a new CLI plugin > adds when experimental cli has some plugins configured 1`] = `
"[web]
title = "Redwood App"
port = 8910
apiUrl = "/.redwood/functions"
includeEnvironmentVariables = []

[api]
port = 8911

[experimental]
[experimental.cli]
autoInstall = true

[[experimental.cli.plugins]]
package = "@existing-example/some-package-when-cli-has-some-packages-configured"
[experimental]
[experimental.cli]
[[experimental.cli.plugins]]
package = "@example/test-package-name"
enabled = true
"
`;

exports[`updateTomlConfig > updateTomlConfig configures a new CLI plugin > adds when experimental cli is not configured 1`] = `
"[web]
title = "Redwood App"
port = 8910
apiUrl = "/.redwood/functions"
includeEnvironmentVariables = []

[api]
port = 8911
[experimental]
[experimental.cli]
autoInstall = true

[[experimental.cli.plugins]]
package = "@example/test-package-when-cli-not-configured"
enabled = true
"
`;

exports[`updateTomlConfig > updateTomlConfig configures a new CLI plugin > adds when experimental cli is setup but has no plugins configured 1`] = `
"[web]
title = "Redwood App"
port = 8910
apiUrl = "/.redwood/functions"
includeEnvironmentVariables = []

[api]
port = 8911

[experimental]
[experimental.cli]
autoInstall = true
[experimental]
[experimental.cli]
[[experimental.cli.plugins]]
package = "@example/test-package-when-no-plugins-configured"
enabled = true
"
`;

exports[`updateTomlConfig > updateTomlConfig configures a new CLI plugin > does not add duplicate place when experimental cli has that plugin configured 1`] = `
"[web]
title = "Redwood App"
port = 8910
apiUrl = "/.redwood/functions"
includeEnvironmentVariables = []

[api]
port = 8911

[experimental]
[experimental.cli]
autoInstall = true

[[experimental.cli.plugins]]
package = "@existing-example/some-package-name-already-exists"
"
`;
